clippy_lints/src/if_let_redundant_pattern_matching.rs

```diff
@@ -57,46 +58,121 @@ impl LintPass for Pass {
 
 impl<'a, 'tcx> LateLintPass<'a, 'tcx> for Pass {
     fn check_expr(&mut self, cx: &LateContext<'a, 'tcx>, expr: &'tcx Expr) {
-        if let ExprKind::Match(ref op, ref arms, MatchSource::IfLetDesugar { .. }) = expr.node {
-            if arms[0].pats.len() == 1 {
-                let good_method = match arms[0].pats[0].node {
-                    PatKind::TupleStruct(ref path, ref pats, _) if pats.len() == 1 => {
-                        if let PatKind::Wild = pats[0].node {
-                            if match_qpath(path, &paths::RESULT_OK) {
-                                "is_ok()"
-                            } else if match_qpath(path, &paths::RESULT_ERR) {
-                                "is_err()"
-                            } else if match_qpath(path, &paths::OPTION_SOME) {
-                                "is_some()"
-                            } else {
-                                return;
-                            }
-                        } else {
-                            return;
+        if let ExprKind::Match(ref op, ref arms, ref match_source) = expr.node {
+            let good_method = match match_source {
+                MatchSource::Normal => {
+                    if arms.len() == 2 {
```

Comment:
I think it would be nice to extract the match bodies of `MatchSource::Normal` and `MatchSource::IfLetDesugar` into two separate methods (maybe `find_sugg_for_match` and `find_sugg_for_if_let`). It would help readability and remove some indentation, too.

Reply:
@phansch Thanks for the suggestion! I'll try to separate them into separate methods!

I have one more question! Current lint rule name is `if_let_redundant_pattern_matching` but, after this implementation, it will also lint the normal `match`. Should I rename the lint rule in this PR? Or, it's fine to send a follow-up PR to rename the rule? Thanks!

Reply:
@CYBAI Good point regarding the name. As per the Clippy RFC, we should probably not rename the lint directly. So we have two options:

1. Deprecating `if_let_redundant_pattern_matching` and adding a new lint `redundant_pattern_matching` that handles both `match` and `if let`. Or..
2. Keeping `if_let_redundant_pattern_matching` and adding a separate lint, called `match_redundant_pattern_matching` that handles just the `match`

I think since we're still pre 1.0, we can go with option 1 without problems. We can mention the new lint name in the deprecation message.

@oli-obk Do you prefer any option over the other? Maybe I'm also just overthinking this :thinking:

Reply:
yea, we should do 1. to keep with the RFC.

Reply:
@oli-obk @phansch Yes! I'll rename it as option 1! Thanks for the suggestion!

Reply:
Fixed in 06cd9c5